### PR TITLE
HTTP 500s sent via gRPC are now sent as gRPC errors, with the body as a tailer.

### LINKED ISSUE #35: HTTPoGRPC masking errors
As long as all the IO works, all HTTP responses will be returned as `success`.  We should ensure we return HTTP 500s as `error`.

### PR DESCRIPTION
**Pull request by tomwilkie**
Also, don't need to connect on a background goroutine any more, this is fixed in https://github.com/grpc/grpc-go/issues/976.

Fixes #35

### CODE REVIEW

File: httpgrpc/httpgrpc.go [hunk 1]
```diff
@@ -161,6 +164,40 @@ func (c *Client) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+	}
+
+	return status.ErrorProto(&spb.Status{
+		Code:    resp.Code,
+		Message: string(resp.Body),
+		Details: []*any.Any{a},
+	})
+}
+
+func responseFromError(err error) (*HTTPResponse, bool) {
```
Comment (jml) [inline httpgrpc/httpgrpc.go]: Maybe `httpResponseFromError`?

File: httpgrpc/httpgrpc.go [hunk 2]
```diff
@@ -161,6 +164,40 @@ func (c *Client) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+func errorFromResponse(resp *HTTPResponse) error {
```
Comment (jml) [inline httpgrpc/httpgrpc.go]: Maybe `errorFromHTTPResponse`?

**Comment by bboreham** [comment]
There's [a note in circle.yml](https://github.com/weaveworks/common/blob/master/circle.yml#L14) saying not to change it until this PR is "fixed"; can anyone shed light on what that would mean, and if it has happened?

**Comment by jml** [comment]
No, I don't know, and regret not insisting on more context when I reviewed this.

At a guess, it's related to the CircleCI failure we can see in the commit log:

```
httpgrpc/httpgrpc.go:22:2: cannot find package "google.golang.org/genproto/googleapis/rpc/status" in any of:

	/go/src/github.com/weaveworks/common/vendor/google.golang.org/genproto/googleapis/rpc/status (vendor tree)

	/usr/local/go/src/google.golang.org/genproto/googleapis/rpc/status (from $GOROOT)

	/go/src/google.golang.org/genproto/googleapis/rpc/status (from $GOPATH)

httpgrpc/httpgrpc.go:24:2: cannot find package "google.golang.org/grpc/status" in any of:

	/go/src/github.com/weaveworks/common/vendor/google.golang.org/grpc/status (vendor tree)

	/usr/local/go/src/google.golang.org/grpc/status (from $GOROOT)

	/go/src/google.golang.org/grpc/status (from $GOPATH)
```

I would suggest removing the workaround (i.e. installing `dep` in the normal way, whatever that is), and seeing what happens.

**Pull request merged**